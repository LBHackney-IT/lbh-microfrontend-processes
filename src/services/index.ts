export { default as locale } from "./locale";
export * as processes from "./processes";
