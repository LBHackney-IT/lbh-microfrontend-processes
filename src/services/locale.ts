import reactHtmlParser from "react-html-parser";

import { AssetAddress } from "@mtfh/common/lib/api/asset/v1";

const locale = {
  backButton: "Back",
  cancel: "Cancel",
  selectProcessDescription: () =>
    reactHtmlParser(
      "<p>Start a process by selecting a process category below. <br /> Where a process is not yet supported in <strong>Manage My Home</strong> these links will open the appropriate interim Google form.</p>",
    ),
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this name.",
  },
  title: "Processes",
  loadingText: "Submitting...",
  components: {
    startProcess: {
      buttonLabel: "Start process",
      thirdPartyHeading: "Sharing information with third parties",
    },
    entitySummary: {
      tenurePaymentRef: "Tenure payment ref",
      address: (assetAddress: AssetAddress): string => {
        const {
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        } = assetAddress;

        return [
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,

          addressLine4,
          postCode,
        ]
          .filter((addressLine) => !!addressLine)
          .join(" ");
      },
    },
  },
};

export default locale;
