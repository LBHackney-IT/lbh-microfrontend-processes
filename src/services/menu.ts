import * as processes from "./processes";

interface MenuProps {
  label: string;
  link: string;
  processes?: { label: string; link: string; targetType: string }[];
}

const getProcessLink = (processName: string) => {
  const process = processes[processName];
  return {
    targetType: process.targetType,
    label: process.title,
    link: `/processes/${process.processName}/start`,
  };
};

const menu: MenuProps[] = [
  {
    label: "Changes to a tenancy",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    processes: [getProcessLink(processes.soletojoint.processName)],
  },
  {
    label: "Record a tenancy / audit homecheck",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfyXscVvgCBhbBBe7ADpV6cUqJoKrrTbU0YYRB_uXS2J4AHMQ/viewform",
  },
  {
    label: "Find legal referral response",
    link: "https://docs.google.com/spreadsheets/d/1fd6FwBXvBo2DmX8qZaXUr7hX9sE6Zr7HZsPlA_OO8Hs/edit#gid=1251146140",
  },
  {
    label: "Request a legal referral",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScpHsoJQgq112HIMbW06YAFpggijwQuODJM483HGbX_PMfCMw/viewform",
  },
  {
    label: "Record notice served",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfRSMia4FeT6WPjrmRlsLpOXyJ4tFs3Sv_O0toxUbv_FmIjKw/viewform",
  },
  {
    label: "Rehousing",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScPXSs4-xxTH2N5dUwtfmEl5mpu7kxZ6mHe86Qy_rlDNu-GbA/viewform",
  },
  {
    label: "Person",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScxrWFLF7M7A3B1yzY-LZFXkV0P0qv6igcQ4wEgWHQk8DaUCw/viewform",
  },
  {
    label: "Report Japanese knotweed",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSceU2H-VWi2ewaEiOAkOWULVFU2DoPSen8SRPyJrmQtRpEhfQ/viewform?gxids=7628",
  },
  {
    label: "Report overgrown tree",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfLmlLraDyOfqYhdjkOfkNWWlHUUep05G5n0GN-ewizN0Fbuw/viewform?gxids=7628",
  },
  {
    label: "Report overgrown garden",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSc3mOUUEQhdRiDsS0k9XDWlfGJG0CEFAC2rnC1sYmM-idt-Nw/viewform?gxids=7628",
  },
];

export default menu;
