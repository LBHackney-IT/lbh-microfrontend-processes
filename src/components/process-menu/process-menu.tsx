import { Link as RouterLink } from "react-router-dom";

import { menu } from "../../services";
import { EntityType } from "../../types";

import { Details, Link } from "@mtfh/common/lib/components";
import { useFeatureToggle } from "@mtfh/common/lib/hooks";

const LegacyMenu = ({ items }) => {
  return (
    <ul className="mtfh-processes-menu__list">
      {items.map((item, index) => (
        <li key={index}>
          <Link variant="link" href={item.link} isExternal>
            {item.label}
          </Link>
        </li>
      ))}
    </ul>
  );
};

const Menu = ({ items, id, entityType }) => {
  return (
    <ul className="mtfh-processes-menu__list">
      {items.map((item, index) => (
        <li key={index}>
          {item.processes ? (
            <Details title={item.label}>
              <ul className="mtfh-processes-menu__process-list">
                {item.processes.map((process, pIndex) => (
                  <li key={pIndex}>
                    <Link
                      variant="link"
                      as={RouterLink}
                      to={`${process.link}/${entityType}/${id}`}
                    >
                      {process.label}
                    </Link>
                  </li>
                ))}
              </ul>
            </Details>
          ) : (
            <Link variant="link" isExternal href={item.link}>
              {item.label}
            </Link>
          )}
        </li>
      ))}
    </ul>
  );
};

interface ProcessMenuProps {
  id: string;
  entityType: EntityType;
}

export const ProcessMenu = ({ id, entityType }: ProcessMenuProps) => {
  const hasEnhancedProcessMenu = useFeatureToggle("MMH.EnhancedProcessMenu");

  return hasEnhancedProcessMenu ? (
    <Menu items={menu} id={id} entityType={entityType} />
  ) : (
    <LegacyMenu items={menu} />
  );
};
