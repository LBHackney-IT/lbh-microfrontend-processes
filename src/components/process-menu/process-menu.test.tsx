import { mockActiveTenureV1, render } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { menu, processes } from "../../services";
import { ProcessMenu } from "./process-menu";

import { $configuration } from "@mtfh/common/lib/configuration";

test("it renders legacy process list", async () => {
  const { container } = render(
    <ProcessMenu id={mockActiveTenureV1.id} targetType="tenure" />,
  );
  expect(container).toMatchSnapshot();
});

describe("feature toggle on", () => {
  beforeEach(() => {
    const features = $configuration.getValue();
    $configuration.next({
      MMH: {
        ...features.MMH,
        featureToggles: { EnhancedProcessMenu: true },
      },
    });
  });

  test("it renders expanded process list and redirects to process on click", async () => {
    render(<ProcessMenu id={mockActiveTenureV1.id} targetType="tenure" />);
    const input = screen.getByText(menu[0].label);
    userEvent.click(input);

    const processLink = await screen.findByText(processes.soletojoint.title);

    expect(processLink).toBeInTheDocument();

    processLink.click();

    expect(window.location.pathname).toContain("/processes/soletojoint");
  });

  test("it opens an external link in a new tab", async () => {
    render(<ProcessMenu id={mockActiveTenureV1.id} targetType="tenure" />);

    const processLink = screen.getByText(menu[2].label);

    expect(processLink).toHaveAttribute("href", `${menu[2].link}`);
    expect(processLink).toHaveAttribute("target", "_blank");
  });

  test("it only shows soletojoint for the specified targetType tenure", async () => {
    render(<ProcessMenu id={mockActiveTenureV1.id} targetType="tenure" />);

    expect(screen.queryByText(processes.soletojoint.title)).toBeInTheDocument();
  });

  test("it only shows soletojoint for the specified targetType property", async () => {
    render(<ProcessMenu id={mockActiveTenureV1.id} targetType="property" />);

    expect(screen.queryByText(processes.soletojoint.title)).not.toBeInTheDocument();
  });
});
