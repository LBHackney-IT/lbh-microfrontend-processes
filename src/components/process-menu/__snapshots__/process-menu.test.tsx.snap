// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders legacy process list 1`] = `
<div>
  <ul
    class="mtfh-processes-menu__list"
  >
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform"
        rel="noopener noreferrer"
        target="_blank"
      >
        Changes to a tenancy
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLSfyXscVvgCBhbBBe7ADpV6cUqJoKrrTbU0YYRB_uXS2J4AHMQ/viewform"
        rel="noopener noreferrer"
        target="_blank"
      >
        Record a tenancy / audit homecheck
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/spreadsheets/d/1fd6FwBXvBo2DmX8qZaXUr7hX9sE6Zr7HZsPlA_OO8Hs/edit#gid=1251146140"
        rel="noopener noreferrer"
        target="_blank"
      >
        Find legal referral response
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLScpHsoJQgq112HIMbW06YAFpggijwQuODJM483HGbX_PMfCMw/viewform"
        rel="noopener noreferrer"
        target="_blank"
      >
        Request a legal referral
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLSfRSMia4FeT6WPjrmRlsLpOXyJ4tFs3Sv_O0toxUbv_FmIjKw/viewform"
        rel="noopener noreferrer"
        target="_blank"
      >
        Record notice served
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLScPXSs4-xxTH2N5dUwtfmEl5mpu7kxZ6mHe86Qy_rlDNu-GbA/viewform"
        rel="noopener noreferrer"
        target="_blank"
      >
        Rehousing
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLScxrWFLF7M7A3B1yzY-LZFXkV0P0qv6igcQ4wEgWHQk8DaUCw/viewform"
        rel="noopener noreferrer"
        target="_blank"
      >
        Person
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLSceU2H-VWi2ewaEiOAkOWULVFU2DoPSen8SRPyJrmQtRpEhfQ/viewform?gxids=7628"
        rel="noopener noreferrer"
        target="_blank"
      >
        Report Japanese knotweed
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLSfLmlLraDyOfqYhdjkOfkNWWlHUUep05G5n0GN-ewizN0Fbuw/viewform?gxids=7628"
        rel="noopener noreferrer"
        target="_blank"
      >
        Report overgrown tree
      </a>
    </li>
    <li>
      <a
        class="govuk-link lbh-link"
        href="https://docs.google.com/forms/d/e/1FAIpQLSc3mOUUEQhdRiDsS0k9XDWlfGJG0CEFAC2rnC1sYmM-idt-Nw/viewform?gxids=7628"
        rel="noopener noreferrer"
        target="_blank"
      >
        Report overgrown garden
      </a>
    </li>
  </ul>
</div>
`;
